File: Easy/1. Two Sum/sol.py
# Link: https://leetcode.com/problems/two-sum/description/

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        # Initialize a dictionary to store previously seen numbers and their indices
        prevMap = {}  # val: index

        # Iterate through the list of numbers along with their indices
        for i, n in enumerate(nums):

            # Calculate the difference needed to reach the target from the current number
            diff = target - n

            # Check if the difference exists in the previous map
            if diff in prevMap:

                # If the difference exists, return the indices of the two numbers that sum up to the target
                return [prevMap[diff], i]

            # If the difference doesn't exist, store the current number and its index in the dictionary
            prevMap[n] = i
        
        # If no solution is found, return but we were made sure in the description of the question we are guaranteed there is a solution, technically this return is not required
        return None

# End of Solution class

File: Easy/1. Two Sum/sol.ts
function twoSum(nums: number[], target: number): number[] {
    // Iterate through each number in the array
    for (let i = 0; i < nums.length; i++) {

        // Iterate through the remaining numbers in the array
        for (let j = i + 1; j < nums.length; j++) {

            // Check if the sum of the current number and any other number equals the target
            if (nums[i] + nums[j] === target) {
                
                // Return the indices of the two numbers that sum up to the target
                return [i, j];
            }
        }
    }
    
    // No solution found
    return [-1, -1];
}